src/service/viewer-impl.js

```diff
@@ -232,8 +232,62 @@ export class Viewer {
     // Wait for document to become visible.
     this.docState_.onVisibilityChanged(this.onVisibilityChange_.bind(this));
 
-    /** @const @private {boolean} */
-    this.isTrustedViewer_ = this.calcTrustedViewer_();
+    let trustedViewerResolved;
+    let trustedViewerPromise;
+    if (!this.isEmbedded_) {
+      // Not embedded in IFrame - can't trust the viewer.
+      trustedViewerResolved = false;
+      trustedViewerPromise = Promise.resolve(false);
+    } else if (this.win.location.ancestorOrigins) {
+      // Ancestors when available take precedence. This is the main API used
+      // for this determination. Fallback is only done when this API is not
+      // supported by the browser.
+      trustedViewerResolved = (this.win.location.ancestorOrigins.length > 0 &&
+          this.isTrustedViewerOrigin_(this.win.location.ancestorOrigins[0]));
+      trustedViewerPromise = Promise.resolve(trustedViewerResolved);
+    } else {
+      // Wait for comms channel to confirm the origin.
+      trustedViewerResolved = undefined;
+      trustedViewerPromise = new Promise(resolve => {
+        /** @const @private {!function(boolean)|undefined} */
+        this.trustedViewerResolver_ = resolve;
+      });
+    }
+
+    /** @const @private {!Promise<boolean>} */
+    this.isTrustedViewer_ = trustedViewerPromise;
+
+    /** @private {string} */
+    this.unconfirmedReferrerUrl_ =
+        this.isEmbedded() && this.params_['referrer'] &&
+            trustedViewerResolved !== false ?
```

Comment:
Isn't `trustedViewerResolved !== false` always true if `this.isEmbedded()` is true?